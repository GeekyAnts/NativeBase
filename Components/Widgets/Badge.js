/* @flow */
'use strict';

import React from 'react';
import {View} from 'react-native';
import NativeBaseComponent from '../Base/NativeBaseComponent';
import computeProps from '../../Utils/computeProps';
import Text from './Text';


export default class BadgeNB extends NativeBaseComponent {

    propTypes: {
        style : React.PropTypes.object
    }

    prepareRootProps() {

        var type = {

            backgroundColor:this.props.primary ?
                            this.getTheme().brandPrimary :
                            this.props.success ?
                            this.getTheme().brandSuccess :
                            this.props.info ?
                            this.getTheme().brandInfo :
                            this.props.warning ?
                            this.getTheme().brandWarning :
                            this.props.danger ?
                            this.getTheme().brandDanger :
                            this.getTheme().badgeBg,
            padding: 3,
            paddingHorizontal: 10,
            alignSelf: 'flex-start',
            borderRadius: 13.5,
            height: 27

        }

        var defaultProps = {
            style: type
        }

        return computeProps(this.props, defaultProps);

    }
    render() {
        return(
            <View {...this.prepareRootProps()}>
                <Text style={{ color: (this.props.textStyle && this.props.textStyle.color) ? this.props.textStyle.color : this.getTheme().badgeColor,
                                fontSize: (this.props.textStyle && this.props.textStyle.fontSize) ? this.props.textStyle.fontSize : this.getTheme().fontSizeBase,
                                lineHeight: (this.props.textStyle && this.props.textStyle.lineHeight) ? this.props.textStyle.lineHeight : this.getTheme().lineHeight-1,
                                textAlign: 'center'}}>{this.props.children}
                </Text>
            </View>
        );
    }

}
