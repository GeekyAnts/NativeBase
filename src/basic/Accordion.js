/* eslint-disable react/no-multi-comp */
/* eslint-disable react/prefer-stateless-function */
import React from 'react';
import {
  Animated,
  TouchableWithoutFeedback,
  FlatList,
  StyleSheet,
  View
} from 'react-native';

import variable from '../theme/variables/platform';

import { Text } from './Text';
import { Icon } from './Icon';

class DefaultHeader extends React.Component {
  render() {
    const {
      expanded,
      expandedIcon,
      expandedIconStyle,
      headerStyle,
      icon,
      iconStyle,
      title
    } = this.props;

    const variables = this.context.theme
      ? this.context.theme['@@shoutem.theme/themeStyle'].variables
      : variable;

    return (
      <View
        style={[
          // eslint-disable-next-line no-use-before-define
          styles.defaultHeader,
          headerStyle || { backgroundColor: variables.headerStyle }
        ]}
      >
        <Text> {title}</Text>
        <Icon
          style={[
            { fontSize: variables.accordionIconFontSize },
            expanded
              ? expandedIcon && expandedIconStyle
                ? expandedIconStyle
                : { color: variables.expandedIconStyle }
              : icon && iconStyle
              ? iconStyle
              : { color: variables.iconStyle }
          ]}
          name={
            expanded ? expandedIcon || 'ios-arrow-up' : icon || 'ios-arrow-down'
          }
        />
      </View>
    );
  }
}

class DefaultContent extends React.Component {
  render() {
    const { content, contentStyle } = this.props;
    const variables = this.context.theme
      ? this.context.theme['@@shoutem.theme/themeStyle'].variables
      : variable;
    return (
      <Text
        style={[
          { padding: variable.accordionContentPadding },
          contentStyle || { backgroundColor: variables.contentStyle }
        ]}
      >
        {content}
      </Text>
    );
  }
}

class AccordionSubItem extends React.Component {
  state = {
    fadeAnim: new Animated.Value(0.3)
  };
  componentDidMount() {
    Animated.timing(this.state.fadeAnim, {
      toValue: 1,
      duration: 500,
      useNativeDriver: true
    }).start();
  }
  render() {
    const { children, style } = this.props;
    const { fadeAnim } = this.state;
    return (
      <Animated.View style={{ ...style, opacity: fadeAnim }}>
        {children}
      </Animated.View>
    );
  }
}

class AccordionItem extends React.Component {
  render() {
    const {
      contentStyle,
      expanded,
      expandedIcon,
      expandedIconStyle,
      headerStyle,
      icon,
      iconStyle,
      index,
      item,
      onAccordionClose,
      onAccordionOpen,
      renderContent,
      renderHeader,
      setSelected
    } = this.props;

    return (
      <View>
        <TouchableWithoutFeedback
          onPress={() => {
            onAccordionOpen && !expanded && onAccordionOpen(item, index);
            onAccordionClose && expanded && onAccordionClose(item, index);
            setSelected(index);
          }}
        >
          <View>
            {renderHeader ? (
              renderHeader(item, expanded)
            ) : (
              <DefaultHeader
                expanded={expanded}
                expandedIcon={expandedIcon}
                expandedIconStyle={expandedIconStyle}
                headerStyle={headerStyle}
                icon={icon}
                iconStyle={iconStyle}
                title={item.title}
              />
            )}
          </View>
        </TouchableWithoutFeedback>
        {expanded ? (
          <AccordionSubItem>
            {renderContent ? (
              renderContent(item)
            ) : (
              <DefaultContent
                content={item.content}
                contentStyle={contentStyle}
              />
            )}
          </AccordionSubItem>
        ) : null}
      </View>
    );
  }
}

export class Accordion extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      selected: []
    };
  }

  setSelected(index) {
    const { expandMultiple } = this.props;
    const selected = this.state.selected.slice();

    if (selected.indexOf(index) !== -1) {
      selected.splice(selected.indexOf(index), 1);
      this.setState({ selected });
    } else {
      selected.push(index);
      this.setState({ selected: expandMultiple ? selected : [index] });
    }
  }

  componentDidMount() {
    const { expanded, expandMultiple } = this.props;
    if (expanded !== undefined && expanded !== null) {
      let selected = Array.isArray(expanded) ? expanded : [expanded];
      selected = expandMultiple ? selected : selected.slice(0, 1);
      this.setState({ selected });
    }
  }

  render() {
    const {
      contentStyle,
      dataArray,
      expandedIcon,
      expandedIconStyle,
      headerStyle,
      icon,
      iconStyle,
      onAccordionClose,
      onAccordionOpen,
      renderContent,
      renderHeader,
      style
    } = this.props;

    const variables = this.context.theme
      ? this.context.theme['@@shoutem.theme/themeStyle'].variables
      : variable;
    return (
      <FlatList
        data={dataArray}
        extraData={this.state}
        style={[
          {
            borderColor: variables.accordionBorderColor,
            borderWidth: variables.borderWidth
          },
          style
        ]}
        keyExtractor={(item, index) => String(index)}
        renderItem={({ item, index }) => (
          <AccordionItem
            key={String(index)}
            contentStyle={contentStyle}
            expandedIcon={expandedIcon}
            expandedIconStyle={expandedIconStyle}
            headerStyle={headerStyle}
            icon={icon}
            iconStyle={iconStyle}
            item={item}
            expanded={this.state.selected.indexOf(index) !== -1}
            index={index}
            renderContent={renderContent}
            renderHeader={renderHeader}
            onAccordionOpen={onAccordionOpen}
            onAccordionClose={onAccordionClose}
            setSelected={i => this.setSelected(i)}
          />
        )}
        {...this.props}
      />
    );
  }
}

const styles = StyleSheet.create({
  defaultHeader: {
    flexDirection: 'row',
    padding: variable.accordionContentPadding,
    justifyContent: 'space-between',
    alignItems: 'center'
  }
});
