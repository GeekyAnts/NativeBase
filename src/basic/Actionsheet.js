/* eslint-disable no-use-before-define */
/* eslint-disable radix */

import React, { Component } from 'react';
import {
  ActionSheetIOS,
  Dimensions,
  FlatList,
  Modal,
  Platform,
  TouchableOpacity,
  StyleSheet,
  ViewPropTypes,
} from 'react-native';
import { connectStyle } from 'native-base-shoutem-theme';

import mapPropsToStyleNames from '../utils/mapPropsToStyleNames';
import commonColor, { PLATFORM } from '../theme/variables/commonColor';

import { Text } from './Text';
import { Icon } from './Icon';
import { Left } from './Left';
import { Right } from './Right';
import { Body } from './Body';
import { ListItem } from './ListItem';

class ActionSheetContainer extends Component {
  static show(config, callback) {
    this.actionsheetInstance._root.showActionSheet(config, callback);
  }
  static hide() {
    this.actionsheetInstance._root.hideActionSheet();
  }

  constructor(props) {
    super(props);
    this.state = {
      modalVisible: false,
      items: [],
    };
  }

  componentDidMount() {
    if (!this.props.autoHide && this.props.duration) {
      // eslint-disable-next-line no-console
      console.warn(`It's not recommended to set autoHide false with duration`);
    }
  }

  static actionsheetInstance;

  showActionSheet(config, callback) {
    if (Platform.OS === PLATFORM.IOS) {
      if (typeof config.options[0] === 'object') {
        const options = config.options;
        const filtered = options.map((item) => {
          return item.text;
        });

        const filteredConfig = { ...config, options: filtered };
        ActionSheetIOS.showActionSheetWithOptions(filteredConfig, callback);
      } else {
        ActionSheetIOS.showActionSheetWithOptions(config, callback);
      }
    } else {
      this.setState({
        items: config.options,
        title: config.title,
        message: config.message,
        destructiveButtonIndex: config.destructiveButtonIndex,
        cancelButtonIndex: config.cancelButtonIndex,
        modalVisible: true,
        callback,
        style: config.style,
        fontStyle: config.fontStyle,
      });
    }
  }

  hideActionSheet() {
    this.setState({ modalVisible: false });
  }

  render() {
    return (
      <Modal
        animationType={'fade'}
        transparent
        visible={this.state.modalVisible}
        onRequestClose={() => {
          this.state.callback(this.state.cancelButtonIndex);
          this.setState({ modalVisible: false });
        }}
      >
        <TouchableOpacity
          activeOpacity={1}
          onPress={() => {
            this.state.callback(this.state.cancelButtonIndex);
            this.setState({ modalVisible: false });
          }}
          style={styles.containerTouchable}
        >
          <TouchableOpacity
            activeOpacity={1}
            style={[styles.innerTouchable, this.state.style]}
          >
            {this.state.title ? (
              <Text style={styles.touchableText}>{this.state.title}</Text>
            ) : null}
            <FlatList
              style={[
                styles.flatList,
                { marginTop: this.state.title ? commonColor.marginTop : 0 },
              ]}
              data={this.state.items}
              keyExtractor={(item, index) => String(index)}
              renderItem={({ index, item }) => {
                return typeof this.state.items[0] === 'string' ? (
                  <ListItem
                    onPress={() => {
                      this.state.callback(parseInt(index));
                      this.setState({ modalVisible: false });
                    }}
                    style={styles.listItem}
                  >
                    <Text style={this.state.fontStyle}>{item}</Text>
                  </ListItem>
                ) : (
                  <ListItem
                    onPress={() => {
                      this.state.callback(parseInt(index));
                      this.setState({ modalVisible: false });
                    }}
                    style={[
                      styles.listItem,
                      {
                        height: commonColor.listItemHeight,
                      },
                    ]}
                    icon
                  >
                    <Left>
                      <Icon
                        name={item.icon}
                        type={item.iconType}
                        style={{
                          color: item.iconColor ? item.iconColor : undefined,
                        }}
                      />
                    </Left>
                    <Body style={styles.listItemBody}>
                      <Text style={this.state.fontStyle}>{item.text}</Text>
                    </Body>
                    <Right />
                  </ListItem>
                );
              }}
            />
          </TouchableOpacity>
        </TouchableOpacity>
      </Modal>
    );
  }
}

ActionSheetContainer.propTypes = {
  ...ViewPropTypes,
};

const styles = StyleSheet.create({
  containerTouchable: {
    backgroundColor: commonColor.containerTouchableBackgroundColor,
    flex: 1,
    justifyContent: 'flex-end',
  },
  flatList: {
    marginHorizontal: commonColor.marginHorizontal,
  },
  innerTouchable: {
    backgroundColor: commonColor.innerTouchableBackgroundColor,
    minHeight: commonColor.minHeight,
    maxHeight: Dimensions.get('window').height / 2,
    padding: commonColor.padding,
    elevation: commonColor.elevation,
  },
  listItem: {
    borderColor: commonColor.listItemBorderColor,
    marginLeft: commonColor.marginLeft,
  },
  listItemBody: {
    borderColor: commonColor.listItemBorderColor,
    paddingLeft: commonColor.marginLeft / 2,
  },
  touchableText: {
    color: commonColor.touchableTextColor,
  },
});

const StyledActionSheetContainer = connectStyle(
  'NativeBase.ActionSheetContainer',
  {},
  mapPropsToStyleNames
)(ActionSheetContainer);

export { StyledActionSheetContainer as ActionSheetContainer };
