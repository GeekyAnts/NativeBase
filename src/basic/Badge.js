import React, { Component } from 'react';
import { View, Text } from 'react-native';

import { connectStyle } from '@shoutem/theme';
import variables from '../theme/variables/platform';
import mapPropsToStyleNames from '../Utils/mapPropsToStyleNames';

class Badge extends Component {


  renderChildren() {
    if (typeof this.props.children !== 'string') {
      return this.props.children;
    } else {
      const newChildren = [];
      newChildren.push(<Text key="label" style={[{ paddingHorizontal: 3,color: variables.badgeColor, fontSize: variables.fontSizeBase, lineHeight: variables.lineHeight - 1},this.props.textStyle]}>{this.props.children}</Text>);
      return newChildren;
    }
  }

  render() {
    return (
      <View ref={c => this._root = c} {...this.props}>
        {this.renderChildren()}
      </View>
    );
  }

  }

const childrenType = function childrenType(props, propName, component) {
  let error;
  const prop = props[propName];
  React.Children.forEach(prop, (child) => {
    console.log(child);
    if ((typeof child !== 'string') && (child.type.displayName !== 'Styled(Icon)') && (child.type.displayName !== 'Styled(Text)') && (child.type.displayName !== 'Styled(AnimatedComponent)')) {
      error = new Error(`${component} can have only Text or Icon components`);
    }
  });
  return error;
};
Badge.propTypes = {
  ...View.propTypes,
//   children: childrenType,
  style: React.PropTypes.object,
};

const StyledBadge = connectStyle('NativeBase.Badge', {}, mapPropsToStyleNames)(Badge);

export {
  StyledBadge as Badge,
};
