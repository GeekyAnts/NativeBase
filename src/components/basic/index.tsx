export { ScrollView } from './ScrollView';
export { View } from './View';
export { StatusBar } from './StatusBar';
export { FlatList } from './FlatList';
export { SectionList } from './SectionList';
export { KeyboardAvoidingView } from './KeyboardAvoidingView';
