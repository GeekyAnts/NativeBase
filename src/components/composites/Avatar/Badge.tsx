import React, { memo, forwardRef } from 'react';
import Box from '../../primitives/Box';
import { usePropsResolution } from '../../../hooks/useThemeProps';
import type { IAvatarBadgeProps } from './types';
import { useHasResponsiveProps } from '../../../hooks/useHasResponsiveProps';

const AvatarBadge = memo(
  forwardRef((props: IAvatarBadgeProps, ref?: any) => {
    const resolvedProps = usePropsResolution('AvatarBadge', props);
    //TODO: refactor for responsive prop
    if (useHasResponsiveProps(props)) {
      return null;
    }
    return <Box {...resolvedProps} ref={ref} />;
  })
);

AvatarBadge.displayName = 'AvatarBadge';
export default AvatarBadge as (props: IAvatarBadgeProps, ref?: any) => any;
