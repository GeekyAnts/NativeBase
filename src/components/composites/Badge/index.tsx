import React, { memo, forwardRef } from 'react';
import HStack from '../../primitives/Stack/HStack';
import Box from '../../primitives/Box';
import type { IBadgeProps } from './types';
import { usePropsResolution } from '../../../hooks/useThemeProps';
import { useHasResponsiveProps } from '../../../hooks/useHasResponsiveProps';

const Badge = (
  {
    children,
    startIcon,
    rightIcon,
    leftIcon,
    endIcon,

    ...props
  }: IBadgeProps,
  ref: any
) => {
  const { _icon, _text, ...newProps } = usePropsResolution('Badge', props);

  //TODO: refactor for responsive prop
  if (useHasResponsiveProps(props)) {
    return null;
  }

  if (leftIcon) {
    startIcon = leftIcon;
  }
  if (rightIcon) {
    endIcon = rightIcon;
  }

  if (endIcon && React.isValidElement(endIcon)) {
    endIcon = React.Children.map(
      endIcon,
      (child: JSX.Element, index: number) => {
        return React.cloneElement(child, {
          key: `badge-end-icon-${index}`,
          ..._icon,
          ...child.props,
        });
      }
    );
  }
  if (startIcon && React.isValidElement(startIcon)) {
    startIcon = React.Children.map(
      startIcon,
      (child: JSX.Element, index: number) => {
        return React.cloneElement(child, {
          key: `badge-start-icon-${index}`,
          ..._icon,
          ...child.props,
        });
      }
    );
  }

  return (
    <HStack {...newProps} ref={ref}>
      {startIcon ? startIcon : null}
      <Box _text={_text}>{children}</Box>
      {endIcon ? endIcon : null}
    </HStack>
  );
};

export type IBadgeComponentType = (props: IBadgeProps) => JSX.Element;

export default memo(forwardRef(Badge)) as IBadgeComponentType;
export type { IBadgeProps };
