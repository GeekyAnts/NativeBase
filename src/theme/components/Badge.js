import { Platform } from 'react-native';
import _ from 'lodash';

import variable from './../variables/platform';

export default (variables = variable) => {
  const badgeTheme = {
      '.primary': {
        backgroundColor: variables.btnPrimaryBg,
      },
      '.warning': {
        backgroundColor: variables.btnWarningBg,
      },
      '.info': {
        backgroundColor: variables.btnInfoBg,
      },
      '.success': {
        backgroundColor: variables.btnSuccessBg,
      },
      '.danger': {
        backgroundColor: variables.btnDangerBg,
      },
      'NativeBase.Text': {
        color: variables.badgeColor,
        fontSize: variables.fontSizeBase,
        lineHeight: variables.lineHeight - 1,
        textAlign: 'center',
        paddingHorizontal: 3,
      },
      backgroundColor: variables.badgeBg,
      padding: variables.badgePadding,
      paddingHorizontal: 6,
      alignSelf: 'flex-start',
      borderRadius: 13.5,
      height: 27,
  };
  return badgeTheme;
};
