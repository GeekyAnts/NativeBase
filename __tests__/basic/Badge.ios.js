import "react-native";
import React from "react";
import renderer from "react-test-renderer";
import { Badge } from "./../../src/basic/Badge";

// Note: test renderer must be required after react-native.

jest.mock("Platform", () => {
  const Platform = require.requireActual("Platform");
  Platform.OS = "ios";
  return Platform;
});

it("renders regular badge", () => {
  const tree = renderer.create(<Badge>2</Badge>).toJSON();
  expect(tree).toMatchSnapshot();
});
