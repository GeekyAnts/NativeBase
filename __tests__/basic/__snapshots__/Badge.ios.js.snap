exports[`test renders regular badge 1`] = `
<View
  style={Object {}}
  virtual={undefined}>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Array [
        Object {
          "color": "#fff",
          "fontSize": 15,
          "lineHeight": 19,
          "paddingHorizontal": 3,
        },
        undefined,
      ]
    }>
    2
  </Text>
</View>
`;
